src/commands/watch.rs

```diff
@@ -87,54 +88,81 @@ impl WatchCommand {
         ));
 
         // Prepare RuntimeConfig for Watchexec
-        let mut runtime_config = RuntimeConfig::default();
-        let (spin_cmd, spin_args) = self.generate_command();
-        runtime_config.commands(vec![watchexec::command::Command::Exec {
-            prog: spin_cmd,
-            args: spin_args,
-        }]);
         let app_dir = parent_dir(&self.app)?;
-        runtime_config.pathset([app_dir.clone()]);
-        let filter = WatchFilter::new(
-            app_dir.clone(),
-            self.generate_path_patterns().await?,
-            WatchFilter::default_ignore_patterns(),
-        )?;
+        let filter = Arc::new(Filter::new(self.generate_filter_config().await?)?);
+        let state_machine = StateMachine::new(self.skip_build, self.clear);
+        let state_machine_clone = state_machine.clone();
+        let mut runtime_config = RuntimeConfig::default();
+        runtime_config.pathset([app_dir]);
         runtime_config.command_grouped(true);
-        runtime_config.filterer(Arc::new(filter));
+        runtime_config.filterer(filter.clone());
         runtime_config.action_throttle(Duration::from_millis(self.debounce));
+        runtime_config.commands(vec![watchexec::command::Command::Exec {
+            prog: self.generate_command(),
+            args: vec![],
+        }]);
+        runtime_config.on_pre_spawn(move |prespawn: PreSpawn| {
+            let up_args = self.up_args.clone();
+            let manifest_path = self.app.clone().to_str().unwrap().to_owned();
+            let state_machine = state_machine.clone();
+            async move {
+                // Dynamically modify the command we're running based on the state machine
+                let state = state_machine.get_state();
+                let spin_args = WatchCommand::generate_arguments(state, up_args, manifest_path);
```

Comment:
This is pretty subjective, but you might want to consider making the internal `State` private and exposing just what you need here as e.g. `state_machine.is_building() -> bool`

Reply:
Leaving this one as is if only b/c I make heavy use of `get_state` in the tests.

Reply:
Its not just about removing code. If you add another internal state like `BuildingAndDownloading` or something, this code (possibly) doesn't actually care about that new state; it just cares about some boolean property of that state.

Reply:
(all of this with the caveat that it definitely isn't very important in the current codebase; this is just design navel-gazing)